### DISCUSSION TITLE
Usage of generated type from Prisma/Client

### DISCUSSION THREAD

**Discussion by thomasmorice** [prisma/prisma#8365] [Q&A] [answered]
Hello,

I'm trying to leverage as much as possible the generated types from Prisma Client into my typescript app. I have one issue where if I instantiate a new empty object which implements the interface generated by prisma, typescript want a number for the id, technically since the object is not in the database, ID should be null. 

Unfortunately, my Prisma schema model of that object has an id which is `not null` so that creates a difference between the app and the database in term of type. 

I'm not sure on how I should tackle that 😕 any idea ?

**Comment by ryands17** [comment, marked as answer]
@thomasmorice 👋 

Could you share your use case of where you would want to use this? I could provide an alternative solution if I can understand your use case better 🙂

**Reply by thomasmorice** [reply]
Yes sure :) 

So I have my model `Exercise` which I will reduce to his minimum form just for the sake of simplicity:

```prisma
model Exercise {
  id               Int               @id @default(autoincrement())
  name             String
  definition       String?
}
```

In my next application when I want to instantiate an empty exercise, I use 
`import { Exercise } from "@prisma/client";` so I'll get the structure of an Exercise and It'll help me create that empty object: 

```ts
  const emptyExercise : Exercise = {
    id: null,
    name: "",
    definition: "",
  };
```

I can then use that empty exercise in a form, populate the properties, and then call the API to add it to the database later on. 

The problem is, typescript doesn't allow me to create this empty exercise, because `id: null` is wrong, ID cannot be null, since the interface says that ID should not be null. It makes sense for a database to have an ID not null, but for the app, to create an empty exercise, the ID set to null actually makes sense, this exercise DOES NOT have an ID yet :/ 

I hope I'm clear enough, I would love to get a solution, that would still allow me to leverage @prisma/client auto-generated types. 

Cheers

**Reply by ryands17** [reply]
You can use Prisma's built-in type for your form like this:

```ts
import { Prisma } from '@prisma/client';

type ExerciseForm = Prisma.ExerciseCreateInput
```

And then use this in your form.

**Reply by mrob11** [reply]
@thomasmorice mere hours ago, you asked the same question I was just about to ask myself. Thanks, and thanks to @ryands17 for the quick response too! :)

**Reply by thomasmorice** [reply]
Indeed that was a very quick answer thanks! 

I actually dived a bit deeper and found Partial<Exercice> which could also fit pretty well my case since it accept `undefined` for all the fields, even the required ones :) 

I'm looking at the type Prisma generates also, I saw CreateInputExercice, I need to see if that could work as well. But Omit is another great solution! 

Thanks 😊

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
